src/lib.rs

```diff
@@ -0,0 +1,136 @@
+// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+use std::collections::HashMap;
+use std::hash::Hash;
+use std::marker::PhantomData;
+use std::sync::Mutex;
+
+const HOT_THRESHOLD: u16 = 10;
+
+/// What was the outcome of a `jit_control_point`?
+#[derive(Debug, PartialEq)]
+pub enum ControlOutcome {
+    /// The `jit_control_point` executed a trace: the interpreter does not need to execute a
+    /// trace for this location in the end user's code.
+    ExecutedTrace,
+    /// The `jit_control_point` did not execute a trace: the interpreter needs to execute the
+    /// relevant code for this location in the end user's code.
+    DidntExecuteTrace
+}
+
+/// The current meta-tracing state of a given location in the end-user's code.
+#[derive(Debug)]
+enum LocCount {
+    /// Below `HOT_THRESHOLD`, but counting up towards it.
+    Counting(u16),
+    /// A trace is being collected.
+    Tracing,
+    /// A compiled trace is available.
+    Trace
+}
+
+/// A meta-tracer.
+pub struct MetaTracer<Loc> {
+    counts: Mutex<HashMap<Loc, LocCount>>,
+    phantom: PhantomData<Loc>
+}
+
+impl<Loc: Eq + Hash> MetaTracer<Loc> {
+    pub fn new() -> Self {
+        Self {
+            counts: Mutex::new(HashMap::new()),
+            phantom: PhantomData
+        }
+    }
+
+    /// Attempt to execute a compiled trace for location `loc`: return a `ControlOutcome` to allow
+    /// the end user to determine whether a trace was executed or not.
+    pub fn control_point(&self, loc: Loc) -> ControlOutcome {
+        let mut cnts = self.counts.lock().unwrap();
+        let loc_cnt = cnts.entry(loc).or_insert(LocCount::Counting(0));
+        match *loc_cnt {
+            LocCount::Counting(x) => {
+                if x >= HOT_THRESHOLD {
+                    *loc_cnt = LocCount::Tracing;
+                } else {
+                    *loc_cnt = LocCount::Counting(x + 1);
+                }
+                ControlOutcome::DidntExecuteTrace
+            },
+            LocCount::Tracing => {
+                *loc_cnt = LocCount::Trace;
+                ControlOutcome::DidntExecuteTrace
+            },
+            LocCount::Trace => ControlOutcome::ExecutedTrace
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::sync::Arc;
+    use std::thread;
+    use super::*;
+
+    #[test]
+    fn threshold_passed() {
+        let mt = MetaTracer::new();
+        for _ in 0..HOT_THRESHOLD {
+            assert_eq!(mt.control_point(0), ControlOutcome::DidntExecuteTrace);
+        }
+        assert_eq!(mt.control_point(0), ControlOutcome::DidntExecuteTrace);
+        assert_eq!(mt.control_point(0), ControlOutcome::DidntExecuteTrace);
+        assert_eq!(mt.control_point(0), ControlOutcome::ExecutedTrace);
+    }
+
+    #[test]
+    fn threaded_threshold_passed() {
+        let mt = Arc::new(MetaTracer::new());
```

Comment:
Cool, one of my outstanding questions was "Should we use Arc so that no one thread 'owns' the JIT".